linked_list/linked_list.py

```diff
@@ -2,116 +2,234 @@
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def delete(self, value):
-        pass
+        current = self.head
+        if current is None:
+            return None
+
+        if self.head.value == value:  # check for first
+            self.head = self.head.next
+            return
+
+        while current:
+            if current.next.value != value:
+                current = current.next
+            else:
+                current.next = current.next.next
+                return
 
     # method to print all the values in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
+
     def visit(self):
         helper_list = []
         current = self.head
 
         while current:
             helper_list.append(str(current.value))
             current = current.next
-        
+
         print(", ".join(helper_list))
 
     # method to reverse the singly linked list
     # note: the nodes should be moved and not just the values in the nodes
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def reverse(self):
-        pass
-  
+        if self.head is None:
+            return None
+
+        current = self.head
+        previous = None
+
+        while current is not None:
+            next = current.next # track next node
+            current.next = previous # change next node to previous node/switching
+            previous = current # previous is pointing to something instead of None
+            current = next # current shifts to the next node of the original, which is now previous
+
+        self.head = previous # head now points to what is the former last item
+
+
     ## Advanced/ Exercises
     # returns the value at the middle element in the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n) My thinking is because, even though I index directly, I use the length method O(n)
+    # Space Complexity: O(1)
     def find_middle_value(self):
-        pass
+        list_length = self.length()
+
+        search = list_length // 2
+        
+        return self.get_at_index(search)
 
     # find the nth node from the end and return its value
     # assume indexing starts at 0 while counting to n
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def find_nth_from_end(self, n):
-        pass
+        desired_index = self.length() - n - 1
+        
+        return self.get_at_index(desired_index)
 
     # checks if the linked list has a cycle. A cycle exists if any node in the
     # linked list links to a node already visited.
     # returns true if a cycle is found, false otherwise.
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def has_cycle(self):

```

Comment:
👍  However the time complexity here is O(n^2) because you're using `if current.next in has_visited`.  You could get it down to O(n) using a set.